rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,823 @@
+# RFC 9481 - 2022-02-23 - Configuration Schema
+
+Vector's configuration, while driven by code, is hard to document, as well as programmatically
+describe and validate outside of Vector itself. This RFC proposes an enhancement to how we define
+configuration types so that Vector can emit an authoritative schema that can be used to drive
+configuration editing and validation, in both interactive and programmatic contexts.
+
+## Context
+
+- Document options naming and remediate [#3714][gh_3714]
+- Formalize how options are deprecated [#4023][gh_4023]
+- Versioned Vector configuraton [#6231][gh_6231]
+- Fix broken handling of adaptive_concurrency defaults [#8189][gh_8189]
+- Fuzz/validate Vector's configuration options [#8229][gh_8229]
+- ARC was not enabled by default [#9727][gh_9727]
+- Avoid enum deserialization for configuration [#10582][gh_10582]
+
+## Cross cutting concerns
+
+- This work will eventually help power the UI for Datadog's Observability Pipelines, in terms of
+  validating user-supplied configurations. We want to be cognizant of areas where we can provide
+  more flexibility/generalization to allow that work to extract whatever information it may require.
+- Powering the new website's reference documentation.
```

Comment:
Worth noting being able to export types to TypeScript would be a massive win for developer experience and correctness on the site.

Reply:
Once you export it to TypeScript, you can babel and transpile it, convert it to ES6, and export the source map.

Reply:
Yep.  There appear to be a bunch of existing JS/TS libraries for parsing a JSONSchema document and generating TypeScript types from the schema, so I don't think this will be a problem at all.